### PR TITLE
Feature/dynamic cluster head association

### PR DESCRIPTION
**Pull request by mattiasnilsen**

### CODE REVIEW

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.c [hunk 1]
```diff
@@ -651,28 +669,19 @@ chaos_round(const uint16_t round_number, const uint8_t app_id, const uint8_t* co
...
-        if (rx_header->cluster_id % 2 != node_id % 2) { 
-          COOJA_DEBUG_PRINTF("NO, not my cluster, cluster_id: %d, slot_nbr: %d, round: %d\n", rx_header->cluster_id, rx_header->slot_number, rx_header->round_number); 
-          slot_number++;
-          LEDS_OFF(LEDS_BLUE);
-          continue;
-        } else { 
-          COOJA_DEBUG_PRINTF("YES, my cluster, cluster_id: %d, slot_nbr: %d, round: %d\n", rx_header->cluster_id, rx_header->slot_number, rx_header->round_number);     
-        }
+      if(!IS_SAME_CLUSTER(rx_header->cluster_id)) {
+        //COOJA_DEBUG_PRINTF("NO, not my cluster, cluster_id: %d, slot_nbr: %d, round: %d\n", rx_header->cluster_id, rx_header->slot_number, rx_header->round_number);
```
Comment (mattiasnilsen) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.c]: Remove commented code

Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.c]: remove?

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.c [hunk 2]
```diff
@@ -949,7 +966,7 @@ uint8_t chaos_associate(rtimer_clock_t* t_sfd_actual_rtimer_ptr, uint16_t *round
         rx_status = chaos_rx_slot(&sfd_vht, 0, 0, 1);
         *t_sfd_actual_rtimer_ptr = VHT_TO_RTIMER(sfd_vht);
       #if CHAOS_CLUSTER
-        associated += (rx_status == CHAOS_TXRX_OK) && (rx_header->cluster_id % 2 == node_id % 2);
+        associated += (rx_status == CHAOS_TXRX_OK); // && IS_SAME_CLUSTER(rx_header->cluster_id, node_id);
```
Comment (mattiasnilsen) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.c]: Remove this? It should not be like this

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.h [hunk 3]
```diff
@@ -220,8 +221,9 @@ enum {
 #define INITIATOR_NODE_ID ((uint16_t)INITIATOR_NODE)
 #endif
 
-//#define IS_INITIATOR()              (node_id == INITIATOR_NODE_ID)
-#define IS_INITIATOR()              (node_id == 3 || node_id == 4)
+#ifndef CHAOS_CLUSTER
```
Comment (mattiasnilsen) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.h]: Look into how to define INITIATOR here even if CHAOS_CLUSTER is defined, it is ugly to define it in two separate files

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.c [hunk 4]
```diff
@@ -860,7 +869,15 @@ chaos_round(const uint16_t round_number, const uint8_t app_id, const uint8_t* co
     /* move to next slot */
     slot_number++;
     /* change channel */
+  #if CHAOS_CLUSTER
+    if(IS_CLUSTER_HEAD() && IS_CLUSTER_ROUND() && HAS_CLUSTER_ID()) {
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.c:868]: create trello card to remember IS_CLUSTER_ROUND is always 0 and HOP_CHANNEL_CLUSTER_HEAD is never used?

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.c (hunk 2 above)
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/chaos.c]: Hej kerp

**Review by Tejpbit — APPROVED** [review]

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    } else if(current_state == CHAOS_RX) {
+        delta = merge_lists(tx_payload, rx_payload);
+        consecutive_rx++;
+        if (delta || consecutive_rx == CONSECUTIVE_RECEIVE_THRESHOLD) {
+            next_state = CHAOS_TX;
+        }
+        
+        int node_in_list = index_of(tx_payload->cluster_head_list, tx_payload->cluster_head_count, node_id);
+        if(node_in_list == -1) {
+            // TODO: Check if list is full. Should we use NODE_LIST_LEN? Do we need to insert, not append?
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c]: trello card?

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    invalid_rx_count = 0;
+    restart_threshold = generate_restart_threshold();
+
+    chaos_round(round_count, app_id, (const uint8_t const*)&cluster_data, sizeof(cluster_data), (7*(RTIMER_SECOND/1000)+0*(RTIMER_SECOND/1000)/2) * CLOCK_PHI, JOIN_ROUND_MAX_SLOTS, get_flags_length(), process);
+}
+
+ALWAYS_INLINE static int is_pending(const uint16_t round_count) {
+    COOJA_DEBUG_PRINTF("cluster: is_pending {rc %u, pending: %u\n", round_count, round_count <= 3);
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c]: remove

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    return round_count <= 3;
+}
+
+static node_id_t pick_best_cluster(const node_id_t *cluster_head_list, uint8_t size) {
+    // Should be based on some more complex calculation later.
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c]: trello card?

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    for(i = 0; i < size; ++i) {
+        if (array[i] == value) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static void round_begin_sniffer(chaos_header_t* header){
+  // header->join = !chaos_get_has_node_index() /*&& !is_join_round*/;
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.c]: remove comments

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.h [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#ifndef _CLUSTER_H
+#define _CLUSTER_H
+
+#include "chaos-control.h"
+
+extern const chaos_app_t cluster;
+extern uint8_t is_cluster_service_running;
+
+#ifndef NODE_LIST_LEN
+#define NODE_LIST_LEN 10  //describes how many nodes can become clusterheads.
```
Comment (Tejpbit) [inline a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.h:10]: Nodes should remember if they've been assigned as cluster heads in previous rounds. in order to let more candidates be elected.

Comment (Tejpbit) [reply a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.h:10]: Trello card

File: a2-synchrotron-contiki/core/net/mac/chaos/chaos.c (hunk 4 above)
Comment (mattiasnilsen) [reply a2-synchrotron-contiki/core/net/mac/chaos/chaos.c:868]: Added

File: a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.h (hunk 9 above)
Comment (mattiasnilsen) [reply a2-synchrotron-contiki/core/net/mac/chaos/service/cluster/cluster.h:10]: Added

**Pull request merged**